### ISSUE TITLE
Allow "Accept-Charset" by default, but block it in higher PL

### ISSUE THREAD

**Issue by dune73** [coreruleset/coreruleset#3148]
The deprecated Accept-Charset header is much more frequent than it should be. And while it's a good indicator of suspicious behavior, the are far too many false positive than what we can accept in the default installation.

See #3140 for a discussion and production evidence.

The problem with the banning implementation: We banned the header via adding it to the existing list of restricted http headers that is enforced at PL1. We need to become creative how to introduce different levels of header restriction for different PLs or some similar mechanism. Opinions welcome.

**Comment by azurit** [comment]
One possible (quite straightforward) solution with only 1 additional rule.

crs-setup.conf:
```
#SecAction \
# "id:900250,\
#  phase:1,\
#  nolog,\
#  pass,\
#  t:none,\
#  setvar:'tx.restricted_headers_pl1=/content-encoding/ /proxy/ /lock-token/ /content-range/ /if/ /x-http-method-override/ /x-http-method/ /x-method-override/',\
#  setvar:'tx.restricted_headers_pl2=/accept-charset/'"
```

REQUEST-901-INITIALIZATION.conf:
```
# Default HTTP policy: restricted_headers (rule 900250 in crs-setup.conf)
SecRule &TX:restricted_headers "@eq 0" \
    "id:901165,\
    phase:1,\
    pass,\
    nolog,\
    ver:'OWASP_CRS/4.0.0-rc1',\
    setvar:'tx.restricted_headers_pl1=/content-encoding/ /proxy/ /lock-token/ /content-range/ /if/ /x-http-method-override/ /x-http-method/ /x-method-override/',\
    setvar:'tx.restricted_headers_pl2=/accept-charset/'"
```

REQUEST-920-PROTOCOL-ENFORCEMENT.conf (in `Paranoia Level 2` section):
```
SecRule REQUEST_HEADERS_NAMES "@rx ^.*$" \
    "id:920XXX,\
    phase:1,\
    block,\
    capture,\
    t:none,t:lowercase,\
    msg:'HTTP header is restricted by policy (%{MATCHED_VAR})',\
    logdata:'Restricted header detected: %{MATCHED_VAR}',\
    tag:'application-multi',\
    tag:'language-multi',\
    tag:'platform-multi',\
    tag:'attack-protocol',\
    tag:'paranoia-level/2',\
    tag:'OWASP_CRS',\
    tag:'capec/1000/210/272',\
    tag:'PCI/12.1',\
    ver:'OWASP_CRS/4.0.0-rc1',\
    severity:'CRITICAL',\
    setvar:'tx.header_name_%{tx.0}=/%{tx.0}/',\
    chain"
    SecRule TX:/^header_name_/ "@within %{tx.restricted_headers_pl2}" \
        "setvar:'tx.inbound_anomaly_score_pl2=+%{tx.critical_anomaly_score}'"
```

**Comment by dune73** [comment]
Your 901165 won't work. If you separate the restricted_headers, you will need to initialize them separately as well. Maybe we can do some acrobatics and still perform this in a single rule, but it will be more complicated.

I wonder though, whether this is understandable for our users. Seeing different lists is really complicated and they will not intuitively understand whether the PL1 list applies to them at PL2. Maybe an approach where the basic restricted list is enriched when running at PL2 is easier to digest.

**Comment by azurit** [comment]
> Your 901165 won't work. If you separate the restricted_headers, you will need to initialize them separately as well.

Oh, sorry, you are right.

**Comment by KramNamez** [comment]
I'm still not sure this is a good idea. In which application are the posted UAs legitimate users? How are these FPs and not just evidence of an IoT botnet or similar issues?

Two old smart TV UAs, 2 ancient smartphone UAs and _Thunderbird_? Sure, maybe that last one could be legitimate, but wouldn't make much more sense to just open an issue for Thunderbird, asking them to fix their behavior?

**Comment by theseion** [comment]
How about we introduce categories of headers? Clearly, the list of headers is a mix where we want to exclude entries for different reasons. We could come up with categories like:
- suspicious custom header
- information leakage
- deprecated but active
- deprecated and forbidden

The `Accept-Charset` header would go into the "deprecated but active" category. That way, we could map each category to a PL.

**Comment by azurit** [comment]
Also, we can add different score by category - 'deprecated but active' may have only a warning severity with default score of 3.

**Comment by KramNamez** [comment]
> > Sure, maybe that last one could be legitimate, but wouldn't it make much more sense to just open an issue for Thunderbird, asking them to fix their behavior?
> 
> +1, this would be great!

I've now tested with Thunderbird 102 - it doesn't actually send that header. Plus, it takes a bunch of effort to even get it to navigate anywhere but specific Mozilla-controlled sites. I'm pretty sure that this header is not genuine. If this is actually an FP, then certainly at least the UA is being spoofed. (It might be behavior of the RSS feed fetching, I haven't tested that, but I would not expect that to be the case. It's the same engine underneath.)

**Comment by azurit** [comment]
@KramNamez I detected two cases of this FP from 2 different users, both of them had this header:
`Accept: text/calendar, application/ics, text/plain;q=0.9`
What about some calendar functionality of Thunderbird? Also, have you teste version 102.8.0?

**Comment by KramNamez** [comment]
I tested 102.7.1 and 112 (daily build), because that's what I had available. I did not test the calendar, because I wasn't aware that that would access websites, but of course pulling ics files makes sense. I wonder if that's a plugin that does something strange, rather than a built-in (which would still use the same core as Firefox, and shouldn't send the header).

But I think instead of derailing this convo further, someone who can actually reproduce this behavior should try and report this to Thunderbird.

I remain opposed to this change, especially to lowering the scoring, because it will mean extra effort in reactivating the rule and upping the scoring again.

**Comment by RedXanadu** [comment]
If we *must* break up the forbidden headers into two lists, it would be good to keep it simple. As simple as possible, maybe?

As long as we document why a header is forbidden in one category or the other, it should be easy to understand? IT would be great to document why some of the existing forbidden headers are forbidden, too.

How about something like this (just an idea, not intended to be used for a PR!)

```
# Forbidden request headers.
#
# -- Headers forbidden at all paranoia levels
#
# The following HTTP headers are considered dangerous and are forbidden by
# default in *all* CRS paranoia levels.
#
# Default: /content-encoding/ /proxy/ /lock-token/ /content-range/ /if/ /x-http-method-override/ /x-http-method/ /x-method-override/
#
# /content-encoding/
#     This header is blocked at this level because...DOCUMENT REASONS
# ...
#
# Uncomment this rule to change the default.
#SecAction \
# "id:900250,\
#  phase:1,\
#  nolog,\
#  pass,\
#  t:none,\
#  setvar:'tx.restricted_headers_pl1=/content-encoding/ /proxy/ /lock-token/ /content-range/ /if/ /x-http-method-override/ /x-http-method/ /x-method-override/'"
#
#
# -- Headers forbidden at paranoia level 2 and above
#
# The following HTTP headers may be exploitable and may be deprecated, but they
# may still be used by some genuine clients. Blocking these headers may cause a
# small number of false positives, hence they are only blocked starting from
# paranoia level 2.
#
# Default: /accept-charset/
#
# /accept-charset/
#     This header is blocked at this level because...DOCUMENT REASONS
#
# Uncomment this rule to change the default.
#SecAction \
# "id:900251,\
#  phase:1,\
#  nolog,\
#  pass,\
#  t:none,\
#  setvar:'tx.restricted_headers_pl2=/accept-charset/'"
```

**Comment by theseion** [comment]
So pretty much what @RedXanadu proposed + some documentation about different categories of headers + renaming the variables to maybe `banned_headers` and `restricted_headers`, or `restricted_headers` and `tentatively_restricted_headers`.

**Comment by dune73** [comment]
Yes, we're getting close to a usable position.

I just want to make sure we are describe the four categories you proposed @theseion.

Proposal for the variable names:
* `restricted_headers_basic`
* `restricted_headers_extended`

This has the advantage of not being tied to PLs apparently. On top basic and extended are two terms that correspond with one another. All we would have to do to explain the names would be to tell users that basic means the banned stuff where there is no discussion. And the extended variant adds headers that might still be in use by benign clients.

Other than that, I think @RedXanadu's proposal description is a very good base.

**Comment by RedXanadu** [comment]
A draft PR is ready for comments: #3152

**Comment by theseion** [comment]
PR has already been merged: #3152.
